### PR TITLE
Fix patient update

### PR DESCRIPTION
**Pull request by sfount**
This pull request re-factors the patient update module with some of the latest standards. It also adds end to end tests verifying that the core functionality is not broken (it was developed before this framework existed). 

Notably: 
- Updated for ui-router 
- Modal services 
- Reduced controller 
- Standardised notification system 

---

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

### CODE REVIEW

File: client/src/partials/patients/edit/PatientGroupService.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  };
+
+  service.updateDebtor = updateDebtor;
+  service.updateGroupConfig = updateGroupConfig;
+
+  /**
+   * This function if responsible for configuring and openening a modal to update
+   * debtor groups given a patient and a complete callback
+   *
+   * @params Object     patientEntity
```
Comment (jniles) [inline client/src/partials/patients/edit/PatientGroupService.js]: I just recently found this out: the correct syntax of indicating parameters is `@param`, not `@params`.  Not blocking, but just for future reference.

File: client/src/partials/patients/edit/PatientGroupService.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    var debtorGroupConfig = {
+      templateUrl : 'partials/patients/edit/updateDebtorGroup.tmpl.html',
+      controller : 'UpdateDebtorGroup as UpdateDebtorGroupCtrl',
+      resolve : {
+        patient : function () {
+          return patientEntity;
+        },
+
+        /** @todo investigate passing multiple params through $modal.close to remove callback methods */
+        updateModel : function () {
```
Comment (jniles) [inline client/src/partials/patients/edit/PatientGroupService.js:41]: Why do you need callbacks?  Do promises not work out of the box, or are you just updating old functionality?

File: client/src/partials/patients/edit/edit.html [hunk 3]
```diff
@@ -77,315 +47,307 @@
...
+                      hospital-number
+                      registered-value="::PatientEditCtrl.medical.hospital_no"
                       required>
-                    {{ "FORM.LABELS.FEMALE" | translate }}
-                  </label>  
+                    <span ng-if="patientDetails.$error.hospitalNumber" class="glyphicon glyphicon-remove form-control-feedback"></span>
+                    <span ng-if="patientDetails.$pending.hospitalNumber"  class="glyphicon glyphicon-hourglass form-control-feedback"></span>
+                    <!-- <p ng-if="patientDetails.$pending.hospitalNumber" class="help-block">Checking hospital number is unique...</p> -->
+
+                    <!-- TODO Translate -->
```
Comment (jniles) [inline client/src/partials/patients/edit/edit.html]: Is this still a TODO?

File: client/src/partials/patients/edit/updateDebtorGroup.modal.js [hunk 4]
```diff
@@ -58,6 +58,6 @@ function UpdateDebtorGroup($scope, $uibModalInstance, debtors, patient, updateMo
   viewModel.closeModal = closeModal;
 
   function closeModal() { 
-    $uibModalInstance.dismiss();
+    $uibModalInstance.close(1, 2, 3);
```
Comment (jniles) [inline client/src/partials/patients/edit/updateDebtorGroup.modal.js]: Is this a leftover experiment?

File: client/test/e2e/shared/components/notify.js [hunk 5]
```diff
@@ -6,6 +6,23 @@ module.exports = {
   verify : function verify() {
     expect(element(by.css('[data-bh-growl-notification]')).isPresent()).to.eventually.equal(true);
   },
+
+  hasSuccess : function hasSuccess() {
```
Comment (jniles) [inline client/test/e2e/shared/components/notify.js:10]: Nice improvement.

**Comment by jniles** [comment]
@sfount, I've completed a review.  Everything LGTM - this code can be merged into master as soon as tests pass.

**Comment by jniles** [comment]
Test pass!  LGTM.

File: client/src/partials/patients/edit/PatientGroupService.js (hunk 2 above)
Comment (sfount) [inline client/src/partials/patients/edit/PatientGroupService.js:41]: This is updating very old code (pre-2.X 2.X). It is also used to need to update the view without closing the modal for some reason, who knows why. 

It should work with Promises without a problem in the future.

**Pull request merged**